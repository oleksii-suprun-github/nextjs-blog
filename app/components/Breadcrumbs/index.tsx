'use client';

import Link from 'next/link';
import { usePathname } from 'next/navigation';

function Breadcrumbs() {
  const pathname = usePathname();
  const breadcrumbs = pathname.split('/').slice(1);

  return (
    <div className="breadcrumbs text-sm">
      <ul>
        <li>
          <Link href="/">Home</Link>
        </li>
        {breadcrumbs.map((item, index) => {
          const isActive = breadcrumbs.length - 1 == index;
          const label = item[0].toUpperCase() + item.replaceAll('-', ' ').slice(1);
          const link = (index == 1 && `${breadcrumbs[0]}/${breadcrumbs[1]}`) || item;
          return <li key={index}>{isActive ? label : <Link href={`/${link}`}>{label}</Link>}</li>;
        })}
      </ul>
    </div>
  );
}

export default Breadcrumbs;
