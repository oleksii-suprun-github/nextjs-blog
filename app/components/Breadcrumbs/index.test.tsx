import { render, screen } from '@testing-library/react';
import { describe, it, expect, vi, Mock } from 'vitest';
import { usePathname } from 'next/navigation';
import Breadcrumbs from '.';

// Mock usePathname from next/navigation
vi.mock('next/navigation', () => ({
  usePathname: vi.fn(),
}));

describe('Breadcrumbs Component', () => {
  it('should render correctly and match snapshot', () => {
    (usePathname as Mock).mockReturnValue('/about/team');
    const { asFragment } = render(<Breadcrumbs />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render Home link', () => {
    (usePathname as Mock).mockReturnValue('/about');
    render(<Breadcrumbs />);
    const homeLink = screen.getByRole('link', { name: 'Home' });
    expect(homeLink).toBeInTheDocument();
    expect(homeLink).toHaveAttribute('href', '/');
  });

  it('should render breadcrumbs based on pathname', () => {
    (usePathname as Mock).mockReturnValue('/about/team');
    render(<Breadcrumbs />);

    const breadcrumbLinks = screen.getAllByRole('link');
    expect(breadcrumbLinks).toHaveLength(2);

    expect(breadcrumbLinks[0]).toHaveTextContent('Home');
    expect(breadcrumbLinks[0]).toHaveAttribute('href', '/');

    expect(breadcrumbLinks[1]).toHaveTextContent('About');
    expect(breadcrumbLinks[1]).toHaveAttribute('href', '/about');

    const teamText = screen.getByText('Team');
    expect(teamText).toBeInTheDocument();
  });

  it('should display correct breadcrumb labels', () => {
    (usePathname as Mock).mockReturnValue('/publications/blog');
    render(<Breadcrumbs />);

    const breadcrumbLinks = screen.getAllByRole('link');
    expect(breadcrumbLinks).toHaveLength(2);

    expect(breadcrumbLinks[1]).toHaveTextContent('Publications');
    expect(breadcrumbLinks[1]).toHaveAttribute('href', '/publications');

    const webDevText = screen.getByText('Blog');
    expect(webDevText).toBeInTheDocument();
  });
});
