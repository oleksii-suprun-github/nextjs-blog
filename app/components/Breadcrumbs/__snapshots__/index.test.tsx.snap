// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Breadcrumbs Component > should render correctly and match snapshot 1`] = `
<DocumentFragment>
  <div
    class="breadcrumbs text-sm text-gray-200"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="/about"
        >
          About
        </a>
      </li>
      <li>
        Team
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
